AtReconstruction/AtFitter/OpenKF/kalman_filter/TrackFitterUKF.h

```diff
@@ -0,0 +1,375 @@
+
+   Matrix<DIM_A, SIGMA_DIM_A> m_matSigmaXa{Matrix<DIM_A, SIGMA_DIM_A>::Zero()}; ///< @brief Sigma points matrix
+
+   TrackFitterUKF()
+      : KalmanFilter<DIM_X, DIM_Z>(), m_kappa(3 - DIM_A), m_matQ(Matrix<DIM_V, DIM_V>::Zero()),
+        m_matR(Matrix<DIM_N, DIM_N>::Zero())
+   {
+      // 1. calculate weights
+      updateWeights();
+   }
+
+   ~TrackFitterUKF() {}
+
+   void setKappa(float32_t kappa)
+   {
+      m_kappa = kappa; // Set the kappa parameter for sigma point calculation
+      updateWeights(); // Update the weights based on the new kappa value
+   }
+
+   // This code uses two different conventions for managing noise.
+   // The state vector noise is set in the updateAugmentedStateAndCovariance() method, while
+   // the noise vectors for the process and measurement models are set in the setCovarianceQ() and
+   // setCovarianceR() methods. This is an odd choice. We will be moving everything into a common
+   // structure where updateAugmentedStateAndCovariance() handle all covariance updates that are actually
+   // part of the augmented state vector.
+
+   ///
+   /// @brief adding process noise covariance Q to the augmented state covariance
+   /// matPa in the middle element of the diagonal.
+   ///
+   void setCovarianceQ(const Matrix<DIM_V, DIM_V> &matQ)
+   {
+      m_matQ = matQ; // Store the process noise covariance matrix
+   }
+
+   ///
+   /// @brief set the measurement noise covariance R to be used in the update step
+   ///
+   void setCovarianceR(const Matrix<DIM_N, DIM_N> &matR)
+   {
+      m_matR = matR; // Store the measurement noise covariance matrix
+   }
+
+   /// Add state vector (m_vecX) to the augment state vector (m_vecXa) and also
+   /// add covariance matrix (m_matP) to the augment covariance (m_matPa).
+   void updateAugWithState()
+   {
+      // Copy state vector to augmented state vector
+      for (int32_t i{0}; i < DIM_X; ++i) {
+         m_vecXa[i] = m_vecX[i];
+      }
+
+      // Copy state covariance matrix to augmented covariance matrix
+      for (int32_t i{0}; i < DIM_X; ++i) {
+         for (int32_t j{0}; j < DIM_X; ++j) {
+            m_matPa(i, j) = m_matP(i, j);
+         }
+      }
+   }
+
+   std::array<float32_t, DIM_V> calculateProcessNoiseMean()
+   {
+      // Calculate the expectation value of the process noise using the current value of the state vector m_vecX
+      std::array<float32_t, DIM_V> processNoiseMean{0};
+
+      // TODO: Set the mean energy loss based on the momentum and particle type. Probably best to track stopping power?
+      return processNoiseMean;
+   }
+
+   Matrix<DIM_V, DIM_V> calculateProcessNoiseCovariance()
+   {
+      // Calculate the process noise covariance matrix
+      Matrix<DIM_V, DIM_V> matQ{Matrix<DIM_V, DIM_V>::Zero()};
+      matQ = m_matQ; // Use the stored process noise covariance matrix
+
+      // TODO: Set the process noise covariance for angular straggle and energy loss.
+      return matQ;
+   }
```

Comment:
@Yassid This is what builds the covariance matrix of the process noise. So energy and angular straggeling